pinot-core/src/main/java/org/apache/pinot/core/operator/dociditerators/AndDocIdIterator.java

```diff
@@ -102,6 +102,17 @@ public int next() {
           i = -1;
         }
       }
+      if (hasScanBasedIterators && i == docIdIterators.length - 1) {
+        // this means we found the docId common to all nonScanBased iterators, now we need to ensure
+        // that its also found in scanBasedIterator, if not matched, we restart the intersection
+        for (ScanBasedDocIdIterator iterator : scanBasedDocIdIterators) {
+          if (!iterator.isMatch(currentMax)) {
+            i = -1;
```

Comment:
Can you restructure the (outer) loop so that it is a while with some condition rather than a `for(i = ..)` in which we change the loop counter? I understand it was already that way.

Reply:
@mcvsubbu Planning to do some clean-up in the following pr. Wanted to keep this pr small and focus on the bug fix.